pkg/bootstrap/platform/gcp.go

```diff
@@ -211,3 +228,75 @@ func (e *gcpEnv) Locality() *core.Locality {
 
 	return &l
 }
+
+var (
+	mdOnce   sync.Once
+	project  string
+	zone     string
+	instance string
+	cluster  string
+)
+
+// Retrieves needed metadata strictly from the Google Compute API
+func getMetadataAttributes(e *gcpEnv) {
```

Comment:
I'm curious about why we didn't just use `e.Metadata()` instead of adding this new function?

We could extend `gcpEnv` to hold a `metadata map[string]string` that could be populated via a `sync.Once` (or some other sync) and then we could avoid double lookups of the project, location, instance, and cluster info, right?

I think that is better than introducing package-level vars to hold the temp info (especially since these are properties of the gcpEnv).

Reply:
I was worried about inconsistencies between the environment variables and the ones that the compute platform provides (since e.Metadata uses env variables first). I can move the logic into Metadata but I thought keeping a separate function/set of vars for the compute only variables would be easier to understand.

Reply:
@JimmyCYJ what is setting the GCP_METADATA env var? Is it possible that it will differ from the information returned from the metadata server?

@justinwei2 I do think it is important to capture this information in `gcpEnv` and not at the package level. So, if @JimmyCYJ responds that there will/could be drift, we should add vars to the `gcpEnv` struct to hold that information and modify this function to have a receiver of `gcpEnv`. Then we can have `Metadata` call this new function to set the values before population. Does that sound OK?

Reply:
I think the problem was the metadata server was flaky, so pods would permanently get bad info. adding GCP_METADATA was a hack to basicaly make this info static

Reply:
@douglas-reid I changed the structure of gcpEnv to hold the metadata instead of the functions and added a cache wrapper around it.